src/objective-c/GRPCClient/GRPCInterceptor.m

```diff
@@ -19,146 +19,273 @@
+                  format:@"Interceptor manager must have factories"];
+    }
+    _thisInterceptor = [factories[0] createInterceptorWithManager:self];
+    if (_thisInterceptor == nil) {
+      return nil;
+    }
+    _previousInterceptor = previousInterceptor;
+    _factories = factories;
+    // Generate interceptor
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000 || __MAC_OS_X_VERSION_MAX_ALLOWED >= 101300
+    if (@available(iOS 8.0, macOS 10.10, *)) {
+      _dispatchQueue = dispatch_queue_create(
+          NULL,
+          dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_DEFAULT, 0));
+    } else {
+#else
+    {
+#endif
+      _dispatchQueue = dispatch_queue_create(NULL, DISPATCH_QUEUE_SERIAL);
+    }
+    dispatch_set_target_queue(_dispatchQueue, _thisInterceptor.dispatchQueue);
+    _transportId = transportId;
   }
-
   return self;
 }
 
-- (void)setPreviousInterceptor:(id<GRPCResponseHandler>)previousInterceptor {
-  _previousInterceptor = previousInterceptor;
-}
-
 - (void)shutDown {
   _nextInterceptor = nil;
   _previousInterceptor = nil;
+  _thisInterceptor = nil;
+  _shutDown = YES;
+}
+
+- (void)createNextInterceptor {
+  NSAssert(_nextInterceptor == nil, @"Starting the next interceptor more than once");
+  NSAssert(_factories.count > 0, @"Interceptor manager of transport cannot start next interceptor");
+  if (_nextInterceptor != nil) {
+    NSLog(@"Starting the next interceptor more than once");
+    return;
+  }
+  NSMutableArray<id<GRPCInterceptorFactory>> *interceptorFactories = [NSMutableArray
+      arrayWithArray:[_factories subarrayWithRange:NSMakeRange(1, _factories.count - 1)]];
+  while (_nextInterceptor == nil) {
+    if (interceptorFactories.count == 0) {
+      _nextInterceptor =
+          [[GRPCTransportManager alloc] initWithTransportId:_transportId previousInterceptor:self];
+      break;
+    } else {
+      _nextInterceptor = [[GRPCInterceptorManager alloc] initWithFactories:interceptorFactories
+                                                       previousInterceptor:self
+                                                               transportId:_transportId];
+      if (_nextInterceptor == nil) {
+        [interceptorFactories removeObjectAtIndex:0];
+      }
+    }
+  }
+  NSAssert(_nextInterceptor != nil, @"Failed to create interceptor or transport.");
+  if (_nextInterceptor == nil) {
+    NSLog(@"Failed to create interceptor or transport.");
+  }
 }
 
 - (void)startNextInterceptorWithRequest:(GRPCRequestOptions *)requestOptions
                             callOptions:(GRPCCallOptions *)callOptions {
-  if (_nextInterceptor != nil) {
-    id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
-    dispatch_async(copiedNextInterceptor.requestDispatchQueue, ^{
-      [copiedNextInterceptor startWithRequestOptions:requestOptions callOptions:callOptions];
-    });
+  if (_nextInterceptor == nil && !_shutDown) {
```

Comment:
Do we need locking around `_nextInterceptor` and `_shutDown` to handle the following case?
Thread1
`shutdown()` is called, `_nextInterceptor` is set to nil, but `_shutDown` is not yet set to YES

Thread 2
Running `startNextInterceptorWithRequest`: `_nextInterceptor == nil && !_shutDown`, so it creates a new interceptor.